src/kirin/rewrite/inline.py

```diff
@@ -68,6 +68,112 @@ def inline_call_like(
             args (tuple[ir.SSAValue, ...]): the arguments of the call (first one is the callee)
             region (ir.Region): the region of the callee
         """
+        if not call_like.parent_block:
+            return
+
+        if not call_like.parent_region:
+            return
+
+        # NOTE: we cannot change region because it may be used elsewhere
+        inline_region: ir.Region = region.clone()
+
+        if self._can_use_simple_inline(inline_region):
+            return self._inline_simple(call_like, args, inline_region.blocks[0])
+
+        return self._inline_complex(call_like, args, inline_region)
+
+    def _can_use_simple_inline(self, inline_region: ir.Region) -> bool:
+        """Check if we can use the fast path for simple single-block inlining.
+
+        Args:
+            inline_region: The cloned region to be inlined
+
+        Returns:
+            True if simple inline is possible (single block with simple return)
+        """
+        if len(inline_region.blocks) != 1:
+            return False
+
+        block = inline_region.blocks[0]
+
+        # Last statement must be a simple return
+        if not block.last_stmt or not isinstance(block.last_stmt, func.Return):
+            return False
+
+        return True
+
+    def _inline_simple(
+        self,
+        call_like: ir.Statement,
+        args: tuple[ir.SSAValue, ...],
+        func_block: ir.Block,
+    ):
+        """Fast path: inline single-block function by splicing statements.
+
+        For simple functions with no control flow, we just clone the function's
+        statements and insert them before the call site.
+        No new blocks are created, no statement parent updates are needed.
+
+        Complexity: O(k) where k = number of statements in function (typically small)
+
+        Args:
+            call_like: The call statement to replace
+            args: Arguments to the call (first is callee, rest are parameters)
+            func_block: The single block from the cloned function region
+        """
+        ssa_map: dict[ir.SSAValue, ir.SSAValue] = {}
+        for func_arg, call_arg in zip(func_block.args, args):
+            ssa_map[func_arg] = call_arg
+            if func_arg.name and call_arg.name is None:
+                call_arg.name = func_arg.name
+
+        for stmt in func_block.stmts:
+            if isinstance(stmt, func.Return):
+                return_value = ssa_map.get(stmt.value, stmt.value)
+
+                if call_like.results:
+                    for call_result in call_like.results:
+                        call_result.replace_by(return_value)
+
+                # Don't insert the return statement itself
+                continue
```

Comment:
You probably want break here?  Although in practice I suppose valid blocks can't have Return before the end.